File: packages/transfer-engine-ts/src/chunk-planner.ts
/**
 * Chunk planning.
 *
 * {@link planChunks} splits a file of a given size into an ordered list of
 * {@link ChunkRef}s that exactly tile `[0, fileSize)` — contiguous, no gaps, no
 * overlaps. Because each chunk carries its absolute byte offset, the receiver
 * can write chunks in any order straight to their final positions, which is
 * what makes parallel sending safe (reassembly is implicit).
 *
 * See section 6.1 of `protocol-spec/wrapdrive-protocol-v1.md`.
 */

import type { ChunkRef } from '@wrapdrive/protocol';

/**
 * Plan the chunks for a file.
 *
 * Postconditions (verified by property tests):
 *  - the chunk lengths sum exactly to `fileSize`;
 *  - the ranges tile `[0, fileSize)` with no gaps and no overlaps;
 *  - indices are zero-based and contiguous, and `offset[i]` equals the sum of
 *    all preceding lengths;
 *  - every chunk except the last has `length === chunkSize`, and the last
 *    chunk's length is in `(0, chunkSize]`;
 *  - `fileSize === 0` yields an empty array.
 *
 * @param fileSize  total file size in bytes; must be `>= 0`
 * @param chunkSize maximum chunk size in bytes; must be `> 0`
 * @throws RangeError if `fileSize < 0` or `chunkSize <= 0`
 */
export function planChunks(fileSize: number, chunkSize: number): ChunkRef[] {
  if (!Number.isInteger(fileSize) || fileSize < 0) {
    throw new RangeError('fileSize must be a non-negative integer');
  }
  if (!Number.isInteger(chunkSize) || chunkSize <= 0) {
    throw new RangeError('chunkSize must be a positive integer');
  }

  const chunks: ChunkRef[] = [];
  let offset = 0;
  let index = 0;

  while (offset < fileSize) {
    const length = Math.min(chunkSize, fileSize - offset);
    chunks.push({ index, offset, length });
    offset += length;
    index += 1;
  }

  return chunks;
}

File: apps/android/core/transfer/src/main/kotlin/com/wrapdrive/core/transfer/ChunkPlanner.kt
package com.wrapdrive.core.transfer

import com.wrapdrive.core.protocol.ChunkRef

/**
 * Chunk planning (Kotlin port).
 *
 * Splits a file into an ordered list of [ChunkRef]s that exactly tile
 * `[0, fileSize)` — contiguous, no gaps, no overlaps — identical to the
 * TypeScript `planChunks`. See section 6.1 of the protocol spec.
 */
object ChunkPlanner {
    /**
     * Plan the chunks for a file.
     *
     * @param fileSize total file size in bytes; must be `>= 0`
     * @param chunkSize maximum chunk size in bytes; must be `> 0`
     * @throws IllegalArgumentException if arguments are out of range
     */
    fun planChunks(fileSize: Long, chunkSize: Long): List<ChunkRef> {
        require(fileSize >= 0) { "fileSize must be non-negative" }
        require(chunkSize > 0) { "chunkSize must be positive" }

        val chunks = ArrayList<ChunkRef>()
        var offset = 0L
        var index = 0
        while (offset < fileSize) {
            val length = minOf(chunkSize, fileSize - offset)
            chunks.add(ChunkRef(index = index, offset = offset, length = length))
            offset += length
            index += 1
        }
        return chunks
    }
}
